From e48baa8cf66233298a732d758aa4b37030aea4fc Mon Sep 17 00:00:00 2001
From: Vinu Vaghasia <vinu.vaghasia@amd.com>
Date: Tue, 15 Mar 2022 17:14:47 -0500
Subject: [PATCH 1/1] phosphor-pid-control: Add CPU Temp simulation code

Code added to simulate the CPU Temp form /etc/thermal.d/temp file for
Fan speed testing.

Signed-off-by: Vinu Vaghasia <vinu.vaghasia@amd.com>
---
 pid/zone.cpp | 37 +++++++++++++++++++++++++++++++++++++
 1 file changed, 37 insertions(+)

diff --git a/pid/zone.cpp b/pid/zone.cpp
index 441031a..16be5db 100644
--- a/pid/zone.cpp
+++ b/pid/zone.cpp
@@ -33,6 +33,7 @@
 #include <memory>
 #include <sstream>
 #include <string>
+#include <filesystem>

 using tstamp = std::chrono::high_resolution_clock::time_point;
 using namespace std::literals::chrono_literals;
@@ -357,9 +358,36 @@ void DbusPidZone::updateFanTelemetry(void)
     return;
 }

+static double fileParseTemp(void)
+{
+    std::string errText;
+    std::string fileName = "/etc/thermal.d/temp";
+    try
+    {
+        std::ifstream ifs;
+        ifs.open(fileName);
+        if (ifs)
+        {
+            double value;
+            ifs >> value;
+
+            return value;
+        }
+    }
+    catch (const std::exception& e)
+    {
+        errText = "Exception: ";
+        errText += e.what();
+    }
+
+    return 0;
+}
+
 void DbusPidZone::updateSensors(void)
 {
     using namespace std::chrono;
+    double temp_val;
+
     /* margin and temp are stored as temp */
     tstamp now = high_resolution_clock::now();

@@ -369,6 +397,15 @@ void DbusPidZone::updateSensors(void)
         ReadReturn r = sensor->read();
         int64_t timeout = sensor->getTimeout();

+        /* Following code added for simulating the CPU Temp using /etc/thermal.d/temp file.
+         * If /etc/thermal.d/temp file exists then swampd will read temp value from it
+        */
+        if ( std::filesystem::exists("/etc/thermal.d/temp") )
+        {
+            temp_val = fileParseTemp();
+            r.value = temp_val;
+        }
+
         _cachedValuesByName[t] = r.value;
         tstamp then = r.updated;

--
2.17.1

